docs/end_to_end_tutorials/dev_practices/evaluation.md

```diff
@@ -0,0 +1,26 @@
+# Evaluation
+
+## Setting the Stage
+
+LlamaIndex is meant to connect your data to your LLM applications.
+
+Sometimes, even after diagnosing and fixing bugs by looking at traces, more fine-grained evaluation is required to systematically diagnose issues.
+
+LlamaIndex aims to provide those tools to make identifying issues and receiving useful diagnostic signals easy.
+
+Closely tied to evaluation are the concepts of experimentation and experiment tracking.
+
+While evaluating your pipeline, it is important to keep in mind what parameters your system may be sensitive to.
+
+## Aspects of Evaluation
+```{toctree}
```

Comment:
maybe related to the below point, I think we can be more opinionated about telling users to test in a "bottoms-up" fashion. First test each individual component on its own (e.g. retrieval on its own, the LLM on its own on some ground-truth data), then bit by bit try to tie into an overall system

Reply:
This will also motivate us to be a bit more detailed and specific on things we can help the user with. E.g. show them how to test retrieval in isolation, or the LLM in isolation.

Reply:
I think I'll label this section WIP until we actually ground the use of these targeted evals in a demo example.

Reply:
that's fine, but we can still tell people to test bottoms-up? shouldn't be blocked on us having more demo examples. this provides valuable context

Reply:
> The right procedure IMO is not actually to start with component-wise testing, rather to start with e2e testing to get a sense of how the entire application is doing.
>
> Only when it is verified that the application as a whole works/doesn't work well does it make sense to do more rigorous testing of individual components

Well, I don't actually think that would end up being the recommended way, IMO it's more useful to first start testing e2e and then try to break it down once you start noticing issues.

Reply:
From the e2e eval it would be possible to identify failure cases and add them to an eval set. Then we can optimize against that eval set on specific components (e.g. retrieval, reranking, generation).

Reply:
Hm interesting. That's a fair point in that an e2e flow would directly allow them to identify e2e failure cases.

I'd argue that there's still value in something more bottoms-up though. Because it allows users to sanity check components piece by piece (they work on retrieval itself as an isolated problem, they work on picking the right LLM as an isolated problem), and also gets them to understand each concept more deeply (e.g. understanding IR more deeply).

But I think given your comment I do agree maybe there's not one best practice here. Thoughts on presenting both approaches? And appearing unopinionated

Reply:
e.g. a short paragraph under your above paragraph

Reply:
also actually with a "top-down" approach where you do e2e first, when you run into a failure case it's hard to see which component it's coming from. Starting with bottoms-up you can immediately start sanity-checking each component

Reply:
So what I'm recommending is not to not setup bottoms up eval. 

Rather, in terms of helping the user get to interacting with the system e2e and gaining an intuition quickly, I believe that e2e eval can be a stepping stone into more targeted eval.

Ideally, the user does both. It doesn't really matter which order it is, as long as they get valuable feeback and build it up as they go along. So I guess we can be less strong on the wording.

Ultimately, it's personal preference.

Reply:
We can even draw analogies - component wise eval is a lot closer to unit testing, while e2e testing is closer to integration testing...